### ISSUE TITLE
cost jumps after some epochs

### ISSUE THREAD

**Issue by rpvelloso** [stanfordnlp/GloVe#199]
this happened in different machines, at different points during training. Couldn't find where this is happening in the source code.
Can't share the data :(
 
TRAINING MODEL
Read 563738072 lines.
Initializing parameters...Using random seed 1635360331
done.
vector size: 600
vocab size: 85748
x_max: 100
alpha: 0.75

10/27/21 - 03:59.19PM, iter: 1, cost: 0.071133
10/27/21 - 04:13.09PM, iter: 2, cost: 0.0551224
10/27/21 - 04:27.28PM, iter: 3, cost: 0.0505688
10/27/21 - 04:41.43PM, iter: 4, cost: 0.0454685
10/27/21 - 04:55.38PM, iter: 5, cost: 0.0416584
10/27/21 - 05:09.45PM, iter: 6, cost: 0.0392608
10/27/21 - 05:23.49PM, iter: 7, cost: 0.0376295
10/27/21 - 05:37.49PM, iter: 8, cost: 0.0365679
10/27/21 - 05:51.59PM, iter: 9, cost: 0.0359734
10/27/21 - 06:05.49PM, iter: 10, cost: 34.4493

**Comment by rpvelloso** [comment]
I've made some changes to deal with this issue during training:

TRAINING MODEL
Read 455065004 lines.
Initializing parameters...Using random seed 1635978159
done.
vector size: 600
vocab size: 85748
x_max: 100
alpha: 0.75
epochs: 40
eta: 0.05
11/03/21 - 07:39.33PM, iter: 1, cost: 0.0823551
11/03/21 - 07:56.42PM, iter: 2, cost: 0.064717
11/03/21 - 08:13.47PM, iter: 3, cost: 0.0599455
11/03/21 - 08:30.48PM, iter: 4, cost: 0.0536976
11/03/21 - 08:47.54PM, iter: 5, cost: 0.0487482
11/03/21 - 09:05.05PM, iter: 6, cost: 0.0456012
11/03/21 - 09:22.10PM, iter: 7, cost: 0.0435627
11/03/21 - 09:39.50PM, iter: 8, cost: 0.042235
11/03/21 - 09:56.59PM, iter: 9, cost: 43.8812
11/03/21 - 09:56.59PM cost increased, restoring last training checkpoint and lowering eta from 0.05 to 0.025
11/03/21 - 10:14.01PM, iter: 9, cost: 0.0361162
11/03/21 - 10:31.02PM, iter: 10, cost: 0.033401
11/03/21 - 10:48.04PM, iter: 11, cost: 0.0325761
11/03/21 - 11:05.09PM, iter: 12, cost: 0.0320624


I'm saving grad and weights each epoch, whenever cost increases I restore the last checkpoint and decrease learning rate by a decay rate. Seems to work.

**Comment by rpvelloso** [comment]
another change I've made: changed from double to single precision resulted in a major speed up.
each epoch took 19, now it's about 9min on the same machine.

**Comment by rpvelloso** [comment]
also rewrote it in C++/STL to get rid of malloc's and possible memory leaks.

**Comment by rpvelloso** [comment]
pushed my changes to my own cloned repo https://github.com/rpvelloso/GloVe
1) vocab count & glove rewritten in C++ (will rewrite other modules later);
2) glove runs twice as fast with fp16 math.

closing.
